core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1342,9 +1339,14 @@ private void internalPerformTranslogRecovery(boolean skipTranslogRecovery, boole
             // We set active because we are now writing operations to the engine; this way, if we go idle after some time and become inactive,
             // we still give sync'd flush a chance to run:
             active.set(true);
+            // we have to set it before we recover from the translog as acquring a snapshot from the translog causes a sync which
+            // causes the global checkpoint to be pulled in.
+            globalCheckpointTracker.updateGlobalCheckpointOnReplica(getEngine().getTranslog().getLastSyncedGlobalCheckpoint(),
```

Comment:
same comment as above, here we can call `globalCheckpointTracker.activatePrimaryMode` instead.
This will allow to only leave `activatePrimaryMode` on primary promotion in updateShardState (with exception of 6.x backport which still needs the 5.x backcompat logic for primary relocation activation).